### DISCUSSION TITLE
Direct Unicode Input

### DISCUSSION THREAD

**Discussion by haithium** [kovidgoyal/kitty#7223] [Q&A] [answered]
Is there a way to insert Unicode directly in Kitty without opening this window?

![image](https://github.com/kovidgoyal/kitty/assets/128622475/40759ff6-3a2e-4d24-b27e-784b0be0c272)

**Comment by kovidgoyal** [comment]
Use your systems IME facilities for inputting unicode if you dont want
to use kitty's

**Comment by cybardev** [comment]
Expanding on the maintainer's suggestion...

If that's MacOS, here's a relevant discussion thread: https://discussions.apple.com/thread/1899290

If you're on Ubuntu or a derivative, try `Ctrl + Shift + u`, followed by the unicode number, followed by `Enter`.

If neither, tell us more about your system and we might be able to help. Or consider searching something like "enter unicode character in <your OS/distro/DE/WM>".

**Reply by haithium** [reply]
I'm using **Archlinux**, and yes, `Ctrl+Shift+u` followed by the unicode number inserts the character. But, as I said, in **kitty**, it opens a window to write the unicode, and I want to prevent this. If I unmap `Ctrl+Shift+u` using `map ctrl+shift+u`, it inserts 7;6u, or `map ctrl+shift+u discard_event` and try to write the unicode, **Kitty** inserts the characters I type. I can map the character I want to a keybinding, and it will work as I wanted, but the problem is I'm using **KMonad** to control my keyboard layers, and this will work inside kitty and won't outside it.

**Reply by cybardev** [reply]
Ok, thanks for filling in the missing details.

@kovidgoyal any suggestions on this?

**Reply by kovidgoyal** [reply]
This isnt anything to do with kitty. If the OP doesnt want to use kitty's unicode input facility, its up to them to find and use an alternative. Nothing kitty can do about it.

**Reply by cybardev** [reply]
Kitty _can_ get out of the way and _allow_ users to use the system's unicode input facility. If Kitty would actually do "nothing" when `Ctrl + Shift + u` or `Ctrl + Shift + e` is pressed, then that would likely solve the problem.

The way suggested by @Neurognostic seems to not work for me on Pop!_OS, but maybe @mdSlash can make it work since he's on Arch.

**Reply by haithium** [reply]
> This isnt anything to do with kitty. If the OP doesnt want to use kitty's unicode input facility, its up to them to find and use an alternative. Nothing kitty can do about it.

The entire topic is related to Kitty, and correct me if I'm wrong. Anyway, I opened a new issue #7269 for that.

**Comment by Neurognostic** [comment, marked as answer]
Use IBus' Emoji Input.

https://wiki.archlinux.org/title/IBus#Emoji_input

It will provide an interface for inputting unicode symbols using the symbol's names and aliases for *any* program.

In your `kitty.conf` set the `GLFW_IM_MODULE` environment variable to `ibus`.

```config
env GLFW_IM_MODULE=ibus
```

Then with the default key map press <kbd>super</kbd>+<kbd>.</kbd> to start the input method, type a name like `cat` and press <kbd>space</kbd> to get more results.

**Reply by Neurognostic** [reply]
It is also mentioned here

https://sw.kovidgoyal.net/kitty/glossary/
https://wiki.archlinux.org/title/Kitty#Enable_IME_support

**Reply by haithium** [reply]
Thanks for your help, but the point is that I don't need to use any input UI.

**Reply by Neurognostic** [reply]
How do you expect to insert unicode without any User Interface?

If you mean GUI, there is only a GUI if you open it. Otherwise it inputs the first matched symbol immediately without displaying any GUI.

**Reply by Neurognostic** [reply]
https://github.com/kovidgoyal/kitty/assets/23530447/b2f42b5c-d279-4f7b-8fba-4da8142a6ed0

That's too much UI?

**Reply by haithium** [reply]
No, that's good, how did you do it?

**Reply by Neurognostic** [reply]
I explained it in my original comment, IBus' Emoji Input. 

It's not just emojis either, it works for any unicode symbol; and it works in every program not just Kitty.

**Reply by haithium** [reply]
@Neurognostic Thanks for your help. I'll take a look ASAP.

**Reply by Neurognostic** [reply]
For many language symbols Xcompose is better. But for emojis and other non-language symbols IBus' Emoji input is the most convenient.

**Comment by kovidgoyal** [comment]
On Tue, Mar 26, 2024 at 07:55:44PM -0700, Sheikh wrote:
> Kitty _can_ get out of the way and _allow_ users to use the system's unicode input facility. If Kitty would actually do "nothing" when `Ctrl + Shift + u` is pressed, then that would likely solve the problem.

If you specify a system level shortcut for ctrl+shift+u it will never
even reach kitty. And regardless you can unmap or even discard any
shortcut in kitty as 2 mins reading the manual will tell you.
https://sw.kovidgoyal.net/kitty/conf/#keyboard-shortcuts

**Reply by cybardev** [reply]
You may have missed the OP's 2nd response where they mentioned trying both unmapping and discarding the shortcut to no avail.

Just tried it myself and can confirm that neither work. In zsh, the cursor becomes a block, and in bash, it inserts `7;6u`. Gnome terminal gets the underlined `u` with both shells, letting me insert unicode easily.

**Comment by kovidgoyal** [comment]
On Tue, Mar 26, 2024 at 08:03:50PM -0700, Sheikh wrote:
> You may have missed the OP's 2nd response where they mentioned trying both unmapping and discarding the shortcut to no avail.
> 
> Just tried it myself and can confirm that neither work.

kitty -o 'map ctrl+shift+u discard_event' cat -v

Press ctrl+shift+u observe nothing happens.

**Reply by cybardev** [reply]
Yes. An underlined `u` should appear like it does on Gnome Terminal.

`^[[117;6u` appears for `kitty -o 'map ctrl+shift+u' cat -v`, which is still not the unicode input `u`.

**Comment by kovidgoyal** [comment]
On Tue, Mar 26, 2024 at 08:11:21PM -0700, Sheikh wrote:
> Yes. An underlined `u` should appear like it does on Gnome Terminal.

Where would this underlined u magically appear from? kitty is not GNOME
terminal. It has a different technique for entering unicode characters,
that is far superior to GNOME terminals, since it allows you to search
unicode characters by typing a few letters from their names OR using hex codes.
Underlined you only supports hex codes.

If you want to use some system level technique for doing this then use
an IME or XCompose or similar. If you do, then the ctrl+shift+u or whatever trigger
key your IME uses event wont even reach kitty to become a shortcut.
